### PR TITLE
Make snapshot load faster

### LINKED ISSUE #221: Accelerate snapshot loading
### Description

For Raft node count, snapshot size 12GB, it cost more than 3 min to load snapshot. (about 90 sec to parse data from disk, 90sec to build childrenSet for nodes)
```
2024.03.22 14:57:07.559821 [ 3806684 ] <Information> KeeperSnapshotStore: parseObject from disk
2024.03.22 14:58:37.390746 [ 3806684 ] <Information> KeeperSnapshotStore: build path from nodes
2024.03.22 15:00:08.135964 [ 3806684 ] <Information> KeeperSnapshotStore: Load snapshot done: nodes 58680021, ephemeral nodes 20298, sessions 699, session_id_counter 121406, zxid 52981992779
```

### PR DESCRIPTION
**Pull request by lzydmxy**
### Which issues of this PR fixes:

This PR try to fix #221
For nodes count 58680021, snapshot size 12GB

|                     | Before(sec) | After(sec) |
| ------------------- | ----------- | ---------- |
| Parse from disk     | 90          | 70         |
| Build children path | 90          | 9          |
### Change log:

Optimizations:

1. Reduced one buffer copy and destruction.
2. Storage edges in bucketed based on the parent node's bucket number, and built childrenSet in parallel.
3. Reserved space for childrenSet in advance.

### CODE REVIEW

File: src/Service/KeeperStore.h [hunk 1]
```diff
@@ -297,6 +301,9 @@ class KeeperStore
     /// Build path children after load data from snapshot
     void buildPathChildren(bool from_zk_snapshot = false);
 
+    // Build childrenset for the node in specified block after load data from snapshot.
+    void buildBlockChildren(const std::vector<BlocksEdges> & all_objects_edges, UInt32 block_idx);
```
Comment (JackyWoo) [inline src/Service/KeeperStore.h:305]: from_zk_snapshot is omited

File: src/Service/KeeperStore.h [hunk 2]
```diff
@@ -200,6 +203,7 @@ class KeeperStore
     using SessionIDs = std::vector<int64_t>;
 
     using Watches = std::unordered_map<String /* path, relative of root_path */, SessionIDs>;
+    using BlocksEdges = std::array<Edges, MAP_BLOCK_NUM>;
```
Comment (JackyWoo) [inline src/Service/KeeperStore.h:206]: Add some comment to help understanding

File: src/Service/KeeperStore.h [hunk 3]
```diff
@@ -160,6 +162,7 @@ class ConcurrentMap
     bool erase(String const & key) { return mapFor(key).erase(key); }
 
     InnerMap & mapFor(String const & key) { return maps_[hash_(key) % NumBlocks]; }
+    UInt32 getBlockIndex(String const & key) { return hash_(key) % NumBlocks; }
```
Comment (JackyWoo) [inline src/Service/KeeperStore.h:165]: Please rename block to bucket in whole `InnerMap`

**Comment by JackyWoo** [comment]
@lzydmxy Nice optimization! Please describe your solution.

**Comment by JackyWoo** [comment]
Is there any optimization for parsing data from disk?

**Comment by lzydmxy** [comment]
> @lzydmxy Nice optimization! Please describe your solution.
Here are three optimization points:
Reduced one buffer copy and destruction.
Storage edges in bucketed based on the parent node's bucket number, and built childrenSet in parallel.
Reserved space for childrenSet in advance.

File: src/Service/NuRaftLogSnapshot.cpp [hunk 4]
```diff
@@ -648,11 +650,12 @@ void KeeperSnapshotStore::loadLatestSnapshot(KeeperStore & store)
                             this->objects_path.size());
                         try
                         {
-                            this->parseObject(store, it->second);
+                            this->parseObject(store, it->second, all_objects_edges[obj_idx]);
                         }
                         catch (Exception & e)
                         {
                             LOG_ERROR(log, "parseObject error {}, {}", it->second, getExceptionMessage(e, true));
+                            throw;
```
Comment (JackyWoo) [inline src/Service/NuRaftLogSnapshot.cpp:658]: Please throw `Exception` and add an error code

**Comment by lzydmxy** [comment]
The test failed for apple clang, should we continue support it ?

**Comment by JackyWoo** [comment]
We should not use Apple clang. Let us update the CI env.

**Review by JackyWoo — APPROVED** [review]

**Pull request merged**